src/librustc_mir/borrow_check/mod.rs

```diff
@@ -897,6 +902,53 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                     .enumerate()
                     .for_each(|field| drop_field(self, field));
             }
+
+            // #45696: special-case Box<T> by treating its dtor as
+            // only deep *across owned content*. Namely, we know
+            // dropping a box does not touch data behind any
+            // references it holds (which a Deep Write would touch).
+            ty::TyAdt(def, _) if def.is_box() => {
+                // First, we model the loss of the reference itself
+                // via a shallow write.
+                //
+                // FIXME: Could we just get by with the second step
+                // below? (This step seems like a good thing to do
+                // just as a matter of principle; but none of our
+                // testsuite complained when I tried leaving it out.)
+                self.access_place(
+                    ContextKind::Drop.new(loc),
+                    (drop_place, span),
+                    (ShallowOrDeep::Shallow(None), Write(WriteKind::StorageDeadOrDrop)),
+                    LocalMutationIsAllowed::Yes,
+                    flow_state,
+                );
+
+                // Second, we model the deallocation of the box
+                // contents via a shallow write to a *deref* of the
+                // box, because the shallow write above wouldn't do
+                // such a deref.
+                //
+                // (We use a shallow write because a Deep one would
+                // touch `&mut` references in `T`.)
+                let deref_place = drop_place.clone().deref();
+                self.access_place(
+                    ContextKind::Drop.new(loc),
+                    (&deref_place, span),
+                    (ShallowOrDeep::Shallow(None), Write(WriteKind::StorageDeadOrDrop)),
+                    LocalMutationIsAllowed::Yes,
+                    flow_state,
+                );
+
+                // Third, we model dropping any content owned by the
+                // box by recurring on box contents. This catches
+                // cases like `Box<Box<ScribbleWhenDropped<&mut T>>>`,
+                // while still restricting Write to *owned* content.
+                let ty = erased_drop_place_ty.boxed_ty();
+                debug!("visit_terminator_drop drop-box-content deref_place: {:?} ty: {:?}",
+                       deref_place, ty);
+                self.visit_terminator_drop(loc, term, flow_state, &deref_place, ty, span);
```

Comment:
This looks like unconditional recursion when dropping `struct A(Box<A>, ...)`

Reply:
Note you didn’t put an `Option<Box<A>>` there

Doesn’t the compiler already reject such infinite types? Let me check

Reply:
Hmm okay this may indeed be a problem for me. You can define such types. You won’t be able to make an instance of one, but I imagine one could make a `let a: Option<A> = None;` and then the compiler would possibly infinite loop while trying to deal with the drop of the `Option<A>` ...

Reply:
Note that there are already existing bugs where infinitely sized types can cause the compiler, often specifically dropck, to hang.

Here are a couple (that are still open/unresolved) that I found by skimming over the results when I searched for "recursive" in our issues list: #44933, and #52852

(Plus there is #4287, which I might argue that example falls into in the long term, since our users might like getting a nice warning about the problem.)

Reply:
Also, it turns out that the infinite loop here does allocate memory on each iteration, so we will eventually exhaust our available memory and error out due to that. I don't know whether people think that is better or worse than just diverging execution... :)

Despite the existence of bugs like #44933 and #52852, I will go ahead and try to fix this PR to detect the recursion...

Reply:
Hmm, actually, the existing compiler *already* appears to infinite loop on an example like

```rust
struct A(Box<A>, &'static str);

fn impossible_to_call(a: A) { }
```

(note i'm observing this with AST-borrowck.)

So I'm back to wondering if this is worth trying to fix in this PR. I guess I should strive to avoid introducing new instances of infinite loops...

Reply:
(but I won't be able to include a test, since that would just cause other parts of the compiler to infinite loop...)

Reply:
hmm maybe I was mistaken in my earlier claim that that example currently causes the compiler to infinite loop. I was observing an infinite loop but I'm no longer sure what version of the compiler I was testing at that time. Will investigate...

Reply:
(ah I might have gotten confused because I was also checking the behavior of #52852, so I might have thought the infinite loops I was seeing there also applied to this `struct A(Box<A>);` case...)

Reply:
(fixed, and added corresponding test.)